bulk: use pebble sstable builder to write SSTs

dt opened the pull request:
bulk: add wrapper for pebble SSTable builder to make SSTs
Comparing to the Rocks-based builder in `engine`:

```
name             old time/op    new time/op    delta
WriteSSTable-12    4.66µs ± 6%    3.05µs ±13%   -34.54%  (p=0.000 n=9+10)

name             old speed      new speed      delta
WriteSSTable-12   258MB/s ± 6%   396MB/s ±12%   +53.62%  (p=0.000 n=9+10)

name             old alloc/op   new alloc/op   delta
WriteSSTable-12      300B ± 0%     1662B ± 0%  +453.83%  (p=0.000 n=8+10)

name             old allocs/op  new allocs/op  delta
WriteSSTable-12      0.00           0.00           ~     (all equal)
```

```
name               old time/op    new time/op    delta
ClusterRestore-12    2.09µs ±19%    1.80µs ±10%  -14.05%  (p=0.016 n=5+5)

name               old speed      new speed      delta
ClusterRestore-12  55.5MB/s ±17%  64.4MB/s ±10%  +16.02%  (p=0.016 n=5+5)
```

Release note (performance improvement): speed up file-writing during bulk-ingestion.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37528)
<!-- Reviewable:end -->

danhhz reviewed (commented):
Cool!

Is this also the sstables we're handing to AddSSTable? If so, how much due diligence has been done to make sure the metadata etc matches the old ones? For example, TBI metadata, bloom filters, etc. Does pebble do all of it for us? There really isn't any info in this commit message

Can you also include the `BenchmarkImportWorkload` and `BenchmarkConvertToSSTable` diffs, plz

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-:-LewRVcVA1hqq8MHjCDa:bu557y0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @petermattis)



<!-- Sent from Reviewable.io -->

dt commented:
I haven't done much digging beyond just that the tests passed and running one benchmark. I'm pretty sure the custom collectors are not there, not sure about the prefix bloom filters being the same.

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @dt, and @petermattis)

___
*[pkg/storage/bulk/sst_writer.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LewiCeA99tUGBsdK86y:-LewiCeA99tUGBsdK86z:b-1etah6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L29)):*
> ```Go
> )
> 
> // SSTWriter writes SSTables.
> ```

These are tables used for backup, right? I think that should be called out here as you're using leveldb format tables.
___
*[pkg/storage/bulk/sst_writer.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LewhedO63DgRU8jBWTJ:-LewhedO63DgRU8jBWTK:bwtcgg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L40)):*
> ```Go
> var pebbleOpts = func() *pebbledb.Options {
> 	c := *pebbledb.DefaultComparer
> 	c.Compare = engine.MVCCKeyCompare
> ```

I'd prefer to see all of the fields explicitly populated:

```
c := &pebbledb.Comparer{
  AbbreviatedKey: ...,
  Compare: engine.MVCCKeyCompare,
  Equal: bytes.Equal,
  Split: ...,
  Successor: ...,
  Name: "cockroach_comparator",
}
```

I'm not sure if `DefaultComparer.Successor` works properly for our MVCC keys.
___
*[pkg/storage/bulk/sst_writer.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LewjhVW68qjLwZ0iK6U:-LewjhVW68qjLwZ0iK6V:b-yrg33z) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L41)):*
> ```Go
> 	c := *pebbledb.DefaultComparer
> 	c.Compare = engine.MVCCKeyCompare
> 	opts := &pebbledb.Options{TableFormat: pebbledb.TableFormatLevelDB, Comparer: &c}
> ```

Just want to confirm that you want leveldb format files. I'd only recommend that for "backup" sstables. For every other usage we should be using rocksdb format. 
___
*[pkg/storage/bulk/sst_writer.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Lewj-xsDTpWmKiX-PkX:-Lewj-xsDTpWmKiX-PkY:bbdsjok) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L48)):*
> ```Go
> // MakeSSTWriter creates a new SSTWriter.
> func MakeSSTWriter() SSTWriter {
> 	f := newMemFile(nil)
> ```

I suppose this is evidence that add `vfs.NewMemFile()` would be useful. Also irritating that `sstable.NewWriter` takes a `vfs.File` rather than an `io.Writer`. I think I can fix that for you.
___
*[pkg/storage/bulk/sst_writer.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Lewih_ACviBrI-mxpgi:-Lewih_B7fEuXZr0dZia:b-nuh09l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L63)):*
> ```Go
> 	fw.scratch = engine.EncodeKeyToBuf(fw.scratch[:0], kv.Key)
> 	k := pebbledb.MakeInternalKey(fw.scratch, 0, pebbledb.InternalKeyKindSet)
> 	return fw.fw.Add(k, kv.Value)
> ```

This can be written as `fw.fw.Set(fw.scratch, kv.Value)`.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> I haven't done much digging beyond just that the tests passed and running one benchmark. I'm pretty sure the custom collectors are not there, not sure about the prefix bloom filters being the same.

Those options could have pretty big impacts on users of imported data, so we'll need to dig into them before merging this. Plus probably run some whole system benchmarking

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-:-LewpeoP2PMxGbWZiw2f:b-v08kjj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @dt, and @petermattis)

___
*[pkg/storage/bulk/sst_writer.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LewjhVW68qjLwZ0iK6U:-LewpTQf8YDOKIl68Y9W:br9npol) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L41)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Just want to confirm that you want leveldb format files. I'd only recommend that for "backup" sstables. For every other usage we should be using rocksdb format. 
</blockquote></details>

IIRC everything uses the leveldb format right now. If that's so, I'd prefer we make the change to rockdb format for non-backup uses in a separate PR


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @danhhz, and @dt)

___
*[pkg/storage/bulk/sst_writer.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LewjhVW68qjLwZ0iK6U:-LewqEHIDB0OwCqcd0nE:b-j7q7z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L41)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

IIRC everything uses the leveldb format right now. If that's so, I'd prefer we make the change to rockdb format for non-backup uses in a separate PR
</blockquote></details>

What is "everything"? We certainly use RocksDB format for the sstables in the main database. I hope we use RocksDB format for ingested sstables, though I agree that if we're not currently doing so we should make that change in a separate PR.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr and @petermattis)

___
*[pkg/storage/bulk/sst_writer.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LewiCeA99tUGBsdK86y:-LewlS4P4E0mr4zWq8kK:bppns6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L29)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

These are tables used for backup, right? I think that should be called out here as you're using leveldb format tables.
</blockquote></details>

They are, though actually with another upcoming change, they won't be -- #37496 pushes the iteration down to C++ and uses the rocks writer, so actually after that change, this is used for everything _except backup_.
___
*[pkg/storage/bulk/sst_writer.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LewhedO63DgRU8jBWTJ:-Lewr-m81Pf4GjyZnGW-:b-twtksc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L40)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'd prefer to see all of the fields explicitly populated:

```
c := &pebbledb.Comparer{
  AbbreviatedKey: ...,
  Compare: engine.MVCCKeyCompare,
  Equal: bytes.Equal,
  Split: ...,
  Successor: ...,
  Name: "cockroach_comparator",
}
```

I'm not sure if `DefaultComparer.Successor` works properly for our MVCC keys.

</blockquote></details>

Done (though it wasn't clear to me if I should be doing anything fancier for Successor)
___
*[pkg/storage/bulk/sst_writer.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LewjhVW68qjLwZ0iK6U:-LewkRWg74Fb6LsfnJOp:be3v2us) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L41)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Just want to confirm that you want leveldb format files. I'd only recommend that for "backup" sstables. For every other usage we should be using rocksdb format. 
</blockquote></details>

Currently we always produce LevelDB format files -- even for IMPORT, RESTORE, or index creation -- since the format is set in `DBSstFileWriterNew` without any control by the caller. So currently we're always ingesting Level format. That said, we could easily change that -- in fact, we could always ingest rocks format since we don't actually ingest backup SSTs directly and instead iterate and rewrite them during RESTORE anyway to remap the keys.
___
*[pkg/storage/bulk/sst_writer.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Lewj-xsDTpWmKiX-PkX:-Lewk8wjB1hk6zf-Mo0K:bom1yjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L48)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I suppose this is evidence that add `vfs.NewMemFile()` would be useful. Also irritating that `sstable.NewWriter` takes a `vfs.File` rather than an `io.Writer`. I think I can fix that for you.
</blockquote></details>

I wanted direct access to the `[]byte` which is why I didn't use the MemFS and make a File with it, since to then read the content I'd have to make another copy (same issue we've had on the reading since with LevelDB). 
___
*[pkg/storage/bulk/sst_writer.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Lewih_ACviBrI-mxpgi:-LewqxYm1BImBA0cpdA4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L63)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This can be written as `fw.fw.Set(fw.scratch, kv.Value)`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr and @petermattis)

___
*[pkg/storage/bulk/sst_writer.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LewjhVW68qjLwZ0iK6U:-LewrK8g5ERMsj1vaqZM:b-x037us) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L41)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Currently we always produce LevelDB format files -- even for IMPORT, RESTORE, or index creation -- since the format is set in `DBSstFileWriterNew` without any control by the caller. So currently we're always ingesting Level format. That said, we could easily change that -- in fact, we could always ingest rocks format since we don't actually ingest backup SSTs directly and instead iterate and rewrite them during RESTORE anyway to remap the keys.
</blockquote></details>

But I also concur with @danhhz -- I'd rather swap the implementation, then swap the format, rather than do both at once.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-:-Lf5NqKJ2jL8Qhdp3fVk:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajkr and @dt)

___
*[pkg/storage/bulk/sst_writer.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Lewj-xsDTpWmKiX-PkX:-Lf5NZk309vIlsrOtvBf:bu0jyn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L48)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I wanted direct access to the `[]byte` which is why I didn't use the MemFS and make a File with it, since to then read the content I'd have to make another copy (same issue we've had on the reading since with LevelDB). 
</blockquote></details>

Once https://github.com/petermattis/pebble/pull/141 merges, you can simplify the `memFile` stuff below. `memFileInfo` should no longer be necessary and you'll only need to implement `Write`, `Close`, and `Sync`.
___
*[pkg/storage/bulk/sst_writer.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Lf5NIMK7cvJXym6ZJ6F:-Lf5NIMK7cvJXym6ZJ6G:bnkg848) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a14620af800f0022f988777f747ac09264cbf9e/pkg/storage/bulk/sst_writer.go#L65)):*
> ```Go
> 			return len(k)
> 		}
> 		return keyPartEnd
> ```

Maybe add a comment that this is the same as `enginepb.SplitMVCCKey`.


<!-- Sent from Reviewable.io -->

dt committed:
bulk: minor test helper refactoring

This pulls some of the logic out of an existing test into helpers for:
A) making a big slice of random KV data
B) making an SST from a slice of KV data

These can then be reused in other SST-making tests to follow.

Release note: none.

dt committed:
bulk: add wrapper for pebble SSTable builder to make SSTs

Comparing to the Rocks-based builder in `engine`:

```
name             old time/op    new time/op    delta
WriteSSTable-12    4.66µs ± 6%    3.05µs ±13%   -34.54%  (p=0.000 n=9+10)

name             old speed      new speed      delta
WriteSSTable-12   258MB/s ± 6%   396MB/s ±12%   +53.62%  (p=0.000 n=9+10)

name             old alloc/op   new alloc/op   delta
WriteSSTable-12      300B ± 0%     1662B ± 0%  +453.83%  (p=0.000 n=8+10)

name             old allocs/op  new allocs/op  delta
WriteSSTable-12      0.00           0.00           ~     (all equal)
```

Release note: none.

dt committed:
bulk: switch to pebble-based SSTable builder

```
name               old time/op    new time/op    delta
ClusterRestore-12    2.09µs ±19%    1.80µs ±10%  -14.05%  (p=0.016 n=5+5)

name               old speed      new speed      delta
ClusterRestore-12  55.5MB/s ±17%  64.4MB/s ±10%  +16.02%  (p=0.016 n=5+5)
```

Release note (performance improvement): speed up file-writing during bulk-ingestion.

dt reviewed (commented):
Following up on the concerns @danhhz raised above, I've been thinking about the question of if we were using the same _configuration_ than the RocksDB writer, and how could result in significant -- but potentially hard to debug -- performance differences when later operating on the produced SSTs. 

Talking with @petermattis, we've generally agreed that we trust pebble's own tests to verify pebble creates the SST you ask it to, so this concern is mostly about how we're configuring it, and if that matches how we're configuring the Rocks writer.

"Fortunately", the Rocks-based writer is actually mostly configured from scratch right now and share very little configuration with runtime Rocks config that we've actually tuned. While I'm not actually sure this is a good thin in general, it at least makes it easier to inspect it vs this pebble writer config and at least determine if switching to pebble could be a regression. The prop collectors are about the only thing shared between the rocks writer and the runtime rocksdb.

I chatted with @nvanbenschoten about this a bit, and we're unconvinced that we could really prove much to ourselves by running a workload on it -- even if SSTs were different, the fact they could be compacted away and replaced with rocks-authored SSTs at any time could make it pretty hard to tell. However, if the actual SSTs were exactly the same, _that_ could rule out any behavior differences without doubt.

I talked to @petermattis about this and long-term, we're not sure that we _want_ to constrain pebble to produce byte-for-byte identical SSTs to rocksdb, as that could exclude including different or extra features that we realize we want. However, in the short term, they do happen to be identical, and for the purposes of vetting this change, I think we can test that. If down the road pebble SSTs diverge, we can decide to delete this test, but that that point, I think the test will have done its job: proving that using using the pebble _writer_ didn't change the behavior compared to the rocks _writer_. w.r.t checking that either writer matches runtime-written SSTs, we don't currently have tests for that and this change doesn't really change that (indeed, we know they don't: we explicitly configure them to use a different format).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-:-LlbitaUB8x2XNZLej39:b-staowi)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajkr and @petermattis)

___
*[pkg/storage/bulk/sst_writer.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Lewj-xsDTpWmKiX-PkX:-LlbilKcDviIPZMqh6bw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0f0c3e4cd8c6167988f2a2afa93a0a0a51a64c/pkg/storage/bulk/sst_writer.go#L48)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Once https://github.com/petermattis/pebble/pull/141 merges, you can simplify the `memFile` stuff below. `memFileInfo` should no longer be necessary and you'll only need to implement `Write`, `Close`, and `Sync`.
</blockquote></details>

Done.
___
*[pkg/storage/bulk/sst_writer.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Lf5NIMK7cvJXym6ZJ6F:-LlbhxAc-p0Azr015Yp5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a14620af800f0022f988777f747ac09264cbf9e/pkg/storage/bulk/sst_writer.go#L65)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Maybe add a comment that this is the same as `enginepb.SplitMVCCKey`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis commented:
> However, in the short term, they do happen to be identical, and for the purposes of vetting this change, I think we can test that.

I'm good with this in the short term. The ssts should be byte-for-byte identical. There was a bug last week that was causing that to not be true, but it has been fixed. The one exception to this rule is range-deletion tombstones. Those are stored slightly differently in Pebble, though in a format that is compatible with RocksDB. So if you do add a test around this, be aware of that limitation.

danhhz reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though I didn't look at pebble use closely. exciting! i've wanted this for a while

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-:-Llc6vG48iamLB9n0dOf:b-hmbzuc)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajkr, @dt, and @petermattis)

___
*[pkg/storage/bulk/sst_batcher.go, line 330 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc6irlC2o5YRPUvdXN:-Llc6irlC2o5YRPUvdXO:bgisna6) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d0bf256ed189dae0e27045e059481a2f06096/pkg/storage/bulk/sst_batcher.go#L330)):*
> ```Go
> 				return nil, nil, err
> 			}
> 			w = MakeSSTWriter()
> ```

should SSTWriter have a Reset or something so you can reuse `scratch` and maybe the `memFile`? this also could easily be a followup
___
*[pkg/storage/bulk/sst_writer.go, line 100 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc4K0WCSMx0j16hvKd:-Llc4K0WCSMx0j16hvKe:bo49kn) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d0bf256ed189dae0e27045e059481a2f06096/pkg/storage/bulk/sst_writer.go#L100)):*
> ```Go
> 		binary.BigEndian.PutUint64(minBuf, uint64(t.min.WallTime))
> 	}
> 	if t.min.Logical > 0 {
> ```

is this really the logic? it's contrived, but if there was a timestamp with 0 walltime and non-0 logical, this would encode only the logical time

update: i looked and i don't think this matches the behavior of the c code. it probably produces the same effect in practice, but it's definitely a distraction to go convince myself of that
___
*[pkg/storage/bulk/sst_writer.go, line 129 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc5RuzAc2ejnCqfIWV:-Llc5RuzAc2ejnCqfIWW:b-j8aqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d0bf256ed189dae0e27045e059481a2f06096/pkg/storage/bulk/sst_writer.go#L129)):*
> ```Go
> 
> func (dummyDeleteRangeCollector) Add(key pebble.InternalKey, value []byte) error {
> 	return nil
> ```

panic or return a not implemented error or something? if it does end up being used somehow, it'd be nice if something failed loudly
___
*[pkg/storage/bulk/sst_writer_test.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc3JIA4IyHfV4TL2NF:-Llc3JIA4IyHfV4TL2NG:b115o60) ([raw file](https://github.com/cockroachdb/cockroach/blob/12b1d5c9173933b3a97dce53c9f46990723638f9/pkg/storage/bulk/sst_writer_test.go#L30)):*
> ```Go
> )
> 
> func makeIntTableKVs(t testing.TB, numKeys, valueSize, maxRevisions int) []engine.MVCCKeyValue {
> ```

just to keep this on you radar, i'd love to move these sorts of things to workload so we don't have to keep duplicating just enough of our encoding logic to do stuff like this. https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/workloadccl/format/sstable.go may be relevant. this is not a blocking comment


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajkr, @danhhz, and @petermattis)

___
*[pkg/storage/bulk/sst_batcher.go, line 330 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc6irlC2o5YRPUvdXN:-LlcoSLM2RouGIjvq1wU:b-rrgbgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d0bf256ed189dae0e27045e059481a2f06096/pkg/storage/bulk/sst_batcher.go#L330)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

should SSTWriter have a Reset or something so you can reuse `scratch` and maybe the `memFile`? this also could easily be a followup
</blockquote></details>

Yeah, I think prefer to punt optimizations to followups, since this has the vendor change in it to rebase.

That said, I did take a quick look at doing this, but we can't reuse `memFile`'s backing slice, it is the bytes handed back by `Finish` and `Finish` is not unsafe. Making it unsafe wouldn't buy us much since you can't hand unsafe bytes to DistSender to actually add the SSTable.

So I think best case we'd be avoiding a single kv-sized alloc per-SSTable if we added a reset that could reuse `scratch`. 
___
*[pkg/storage/bulk/sst_writer.go, line 100 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc4K0WCSMx0j16hvKd:-Llcm4-O9WImr3oRSOgo:b-svgyy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d0bf256ed189dae0e27045e059481a2f06096/pkg/storage/bulk/sst_writer.go#L100)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

is this really the logic? it's contrived, but if there was a timestamp with 0 walltime and non-0 logical, this would encode only the logical time

update: i looked and i don't think this matches the behavior of the c code. it probably produces the same effect in practice, but it's definitely a distraction to go convince myself of that
</blockquote></details>

nice catch. The c++ prop collector just compares and tracks the min/max timestamp as raw bytes. I initially wasn't wild about that, but I guess I've come around to it -- it means makes this a) closely match the c++ implementation and b) far smaller, so I'm sold.

In either case I've extended the test to include non-zero logical to ensure it is covered.
___
*[pkg/storage/bulk/sst_writer.go, line 129 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc5RuzAc2ejnCqfIWV:-Llc7Qm17K_XWX_1I7om:bwmb1bx) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d0bf256ed189dae0e27045e059481a2f06096/pkg/storage/bulk/sst_writer.go#L129)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

panic or return a not implemented error or something? if it does end up being used somehow, it'd be nice if something failed loudly
</blockquote></details>

I don't want to do so unconditionally, since it does get passed every key but it just doesn't do anything with them. Added an assertion that the keys are `Sets` and nothing else for now.
___
*[pkg/storage/bulk/sst_writer_test.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc3JIA4IyHfV4TL2NF:-LleC1RWDXPPjY6kyjM8:b-sskh8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/12b1d5c9173933b3a97dce53c9f46990723638f9/pkg/storage/bulk/sst_writer_test.go#L30)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

just to keep this on you radar, i'd love to move these sorts of things to workload so we don't have to keep duplicating just enough of our encoding logic to do stuff like this. https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/workloadccl/format/sstable.go may be relevant. this is not a blocking comment
</blockquote></details>

Yeah, I generally agree, though in this case it isn't quite drop-in since I wanted per-key timestamps and multiple-revisions-per-key. Might be room for something that reusable that isn't quite the normal workload generator stuff though.


<!-- Sent from Reviewable.io -->

danhhz reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37528)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajkr, @danhhz, @dt, and @petermattis)

___
*[pkg/storage/bulk/sst_batcher.go, line 330 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc6irlC2o5YRPUvdXN:-LlgSevRBFSx6WdSQ9_C:bk2llhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d0bf256ed189dae0e27045e059481a2f06096/pkg/storage/bulk/sst_batcher.go#L330)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Yeah, I think prefer to punt optimizations to followups, since this has the vendor change in it to rebase.

That said, I did take a quick look at doing this, but we can't reuse `memFile`'s backing slice, it is the bytes handed back by `Finish` and `Finish` is not unsafe. Making it unsafe wouldn't buy us much since you can't hand unsafe bytes to DistSender to actually add the SSTable.

So I think best case we'd be avoiding a single kv-sized alloc per-SSTable if we added a reset that could reuse `scratch`. 
</blockquote></details>

Ack. I was thinking you could do something with bufalloc.BytesAllocator for memFile, but maybe that's not even a good here. Either way, let's leave them for followups
___
*[pkg/storage/bulk/sst_writer.go, line 100 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-Llc4K0WCSMx0j16hvKd:-LlgTh_A898tgd5rP_Sg:bo7gyo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d0bf256ed189dae0e27045e059481a2f06096/pkg/storage/bulk/sst_writer.go#L100)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nice catch. The c++ prop collector just compares and tracks the min/max timestamp as raw bytes. I initially wasn't wild about that, but I guess I've come around to it -- it means makes this a) closely match the c++ implementation and b) far smaller, so I'm sold.

In either case I've extended the test to include non-zero logical to ensure it is covered.
</blockquote></details>

I was just imagining making the walltime non-zero check into an anything non-zero check and putting the logical inside it. basically just porting the EncodeTimestamp in libroach
```
if t.min != hlc.Timestamp{} {
  minBuf = make([]byte, 8, 12)
  binary.BigEndian.PutUint64(minBuf, uint64(t.min.WallTime))
  if t.min.Logical != 0 {
    minBuf = minBuf[:12]
    binary.BigEndian.PutUint32(minBuf, uint32(t.min.Logical))
  }
}
```
but this is also fine
___
*[pkg/storage/bulk/sst_writer.go, line 96 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LlgTCt05wH3vZwedmHR:-LlgTCt05wH3vZwedmHS:b-cygq2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecd5df98b1fa4cc22f20f0d63206f8de4b98b03e/pkg/storage/bulk/sst_writer.go#L96)):*
> ```Go
> }
> 
> // Name implements
> ```

nit: incomplete
___
*[pkg/storage/bulk/sst_writer.go, line 182 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37528#-LlgTT0hEudRlGVgUKbm:-LlgTT0hEudRlGVgUKbn:b-6503aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecd5df98b1fa4cc22f20f0d63206f8de4b98b03e/pkg/storage/bulk/sst_writer.go#L182)):*
> ```Go
> 	// we already called its Close() in Finish() and we no-op here. Thus the only
> 	// time we expect to be here is in a deferred Close(), in which case the caller
> 	// probably is already returning some other error, so returning on from this
> ```

nit: "returning one from"


<!-- Sent from Reviewable.io -->

dt commented:
TFTRs!

bors r+
